### PR TITLE
add dynamic `from` hints to typespec

### PR DESCRIPTION
**Pull request by ruslandoga**
👋

This PR continues on https://github.com/elixir-ecto/ecto/pull/3592 and adds dynamic hints to typespecs.

### CODE REVIEW

File: lib/ecto/query/builder/from.ex [hunk 1]
```diff
@@ -58,14 +61,14 @@ defmodule Ecto.Query.Builder.From do
   If possible, it does all calculations at compile time to avoid
   runtime work.
   """
-  @spec build(Macro.t(), Macro.Env.t(), atom, String.t | nil, nil | {:ok, String.t | nil} | [String.t]) ::
```
Comment (ruslandoga) [inline lib/ecto/query/builder/from.ex:61]: Why was `{:ok, String.t | nil}` allowed as `hints`?

**Comment by ukutaht** [comment]
Context: `mix dialyzer` fails on https://github.com/plausible/analytics because we use the dynamic hints feature.

File: lib/ecto/query/builder/from.ex [hunk 2]
```diff
@@ -50,6 +50,9 @@ defmodule Ecto.Query.Builder.From do
   end
 
   defp escape_source(query, _env), do: query
+  
+  @typep dynamic_hint :: {atom, term}
+  @typep hints :: String.t() | dynamic_hint | [String.t() | dynamic_hint]
```
[the remark this answers is not in the document]
Comment (josevalim) [reply lib/ecto/query/builder/from.ex:55]: Ignore me, it does not.

File: lib/ecto/query/builder/from.ex [hunk 3]
```diff
@@ -58,14 +61,14 @@ defmodule Ecto.Query.Builder.From do
   If possible, it does all calculations at compile time to avoid
   runtime work.
   """
-  @spec build(Macro.t(), Macro.Env.t(), atom, String.t | nil, nil | {:ok, String.t | nil} | [String.t]) ::
+  @spec build(Macro.t(), Macro.Env.t(), atom, String.t | nil, hints | nil) ::
```
Comment (josevalim) [inline lib/ecto/query/builder/from.ex:64]:
```suggestion
  @spec build(Macro.t(), Macro.Env.t(), atom, {:ok, String.t | nil} | nil, hints | nil) ::
```

File: lib/ecto/query/builder/from.ex (hunk 2 above)
Comment (josevalim) [inline lib/ecto/query/builder/from.ex:55]: Hints is always a list or nil, so we want this:

```suggestion
  @typep hints :: [String.t() | {atom, term}]
```

Comment (josevalim) [inline lib/ecto/query/builder/from.ex:55]: I think we should consider calling `List.wrap` in `Ecto.Query.from` so we always receive a list. This way, we can do this:

```suggestion
  @typep hints :: [String.t() | dynamic_hint]
```

This simplifying both the code and the specs here.

File: lib/ecto/query/builder/from.ex [hunk 4]
```diff
@@ -58,15 +60,13 @@ defmodule Ecto.Query.Builder.From do
   If possible, it does all calculations at compile time to avoid
   runtime work.
   """
-  @spec build(Macro.t(), Macro.Env.t(), atom, String.t | nil, nil | {:ok, String.t | nil} | [String.t]) ::
+  @spec build(Macro.t(), Macro.Env.t(), atom, {:ok, String.t | nil} | nil, hints | nil) ::
```
Comment (josevalim) [inline lib/ecto/query/builder/from.ex:63]:
```suggestion
  @spec build(Macro.t(), Macro.Env.t(), atom, {:ok, String.t | nil} | nil, hints) ::
```

File: lib/ecto/query/builder/from.ex [hunk 5]
```diff
@@ -58,15 +60,13 @@ defmodule Ecto.Query.Builder.From do
...
           {Macro.t(), Keyword.t(), non_neg_integer | nil}
-  def build(query, env, as, prefix, maybe_hints) do
-    hints = List.wrap(maybe_hints)
-
+  def build(query, env, as, prefix, hints) do
     unless Enum.all?(hints, &is_valid_hint/1) do
       Builder.error!(
-        "`hints` must be a compile time string, list of strings, or a tuple " <>
-          "got: `#{Macro.to_string(maybe_hints)}`"
+        "`hints` must be a list of compile time strings or tuples " <>
```
Comment (josevalim) [inline lib/ecto/query/builder/from.ex:68]: The previous warning was correct. We changed the implementation but the user API is the same!

Comment (ruslandoga) [reply lib/ecto/query/builder/from.ex:68]: 🙈

I got carried away.

Comment (ruslandoga) [reply lib/ecto/query/builder/from.ex]: Fixed.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**